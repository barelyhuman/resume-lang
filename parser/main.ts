import { marked } from "marked";

export type ParserOptions = {
  rootDir: string;
  readFile: (p: string) => string;
};

export type Node = {
  __type: "NODE";
  parent?: AST | Node;
  type: string;
  value: unknown;
  children: Node[];
};

export type AST = {
  __type: "ROOT_NODE";
  type: "root";
  children: Node[];
};

const join = (...paths: string[]) => {
  if (paths.find((d) => d == null)) {
    throw new Error("Paths cannot be `undefined`");
  }
  return paths.join("/").replace(/(\.\/)+/g, "./");
};

const DEFAULT_TRANSFORMS = [
  nestedStringLiteralTransform,
  dateTranform,
  urlTransfrom,
];

const KEYWORDS = ["section", "text", "end", "label", "url", "date", "@import"];

function tokenizer(code: string) {
  const tokens = code.split("");
  return tokens;
}

function createTokenTraverser(tokens: string[]) {
  return {
    tokens,
    pos: -1,
    value() {
      return this.tokens[this.pos];
    },
    next(count = 1) {
      this.pos += count;
      return this.tokens[this.pos];
    },
    prev(count = 1) {
      this.pos -= count;
      return this.tokens[this.pos];
    },
    peekPrev(count = 1) {
      return this.tokens[this.pos - count];
    },
    peekNext(count = 1) {
      return this.tokens[this.pos + count];
    },
  };
}

function createNode<T>(type: string, value: T, parent?: AST | Node): Node {
  return {
    __type: "NODE",
    type,
    value,
    children: [],
    parent,
  };
}

function toAst(tokens: string[], parseOptions: ParserOptions) {
  const collector = [];
  const traverse = createTokenTraverser(tokens);
  const ast: AST = {
    __type: "ROOT_NODE",
    type: "root",
    children: [],
  };
  try {
    let astPointer: AST | Node = ast;
    while (traverse.next()) {
      const literal = collector.concat(traverse.value()).join("");
      if (KEYWORDS.includes(literal)) {
        switch (literal) {
          case "section": {
            let sectionId = "";
            for (;;) {
              traverse.next();
              if (!traverse.value()) break;
              if (traverse.value() === "\n") break;
              sectionId += traverse.value();
            }
            const nodeDef: Node = createNode(
              "section",
              sectionId.trim(),
              astPointer
            );
            (astPointer.children ||= []).push(nodeDef);

            collector.length = 0;
            astPointer = nodeDef;
            break;
          }
          case "@import": {
            let importPath = "";
            for (;;) {
              traverse.next();
              if (!traverse.value()) break;
              if (traverse.value() === "\n") break;
              importPath += traverse.value();
            }

            if (!importPath.trim().startsWith('"')) {
              throw new Error(
                "`@imports` must be defined in single or double `\",'` quotes"
              );
            }

            const normalizeImportPath =
              nestedStringLiteralTransform(importPath).replace(
                /\.resume$/,
                ""
              ) + ".resume";

            const secondary = parseOptions.readFile(
              join(parseOptions.rootDir, normalizeImportPath)
            );
            const tree = parse(secondary, parseOptions);
            astPointer.children.push(...tree.children);
            collector.length = 0;
            break;
          }
          case "end": {
            if (astPointer.__type !== "ROOT_NODE") {
              astPointer = astPointer.parent!;
            }
            collector.length = 0;
            break;
          }
          case "text": {
            let textId = "";
            for (;;) {
              traverse.next();
              if (!traverse.value()) break;
              if (traverse.value() === ":") break;
              textId += traverse.value();
            }
            const textValue: string[] = [];
            const posBeforeParse = traverse.pos;
            for (;;) {
              traverse.next();
              if (!traverse.value()) break;
              textValue.push(traverse.value());
            }

            let lookingForEndAt;
            outer: for (
              lookingForEndAt = 0;
              lookingForEndAt < textValue.length;
              lookingForEndAt += 1
            ) {
              let currentPos = lookingForEndAt;
              const token = textValue[lookingForEndAt];
              if (token === "\n") {
                const walker = [];
                inner: for (;;) {
                  currentPos += 1;
                  walker.push(textValue[currentPos]);
                  if (
                    !textValue[currentPos] ||
                    textValue[currentPos] === "\n"
                  ) {
                    if (walker.join("").trim() === "end") {
                      lookingForEndAt += currentPos - lookingForEndAt;
                      break outer;
                    }
                    break inner;
                  }
                }
              }
            }
            // was already at `posBeforeParse`, then moved up till the nearest `end` which was
            // found at lookingForEndAt, we need to move back just before the last `end`
            traverse.pos = posBeforeParse + lookingForEndAt;
            const asString = textValue
              .slice(0, lookingForEndAt - "end".length)
              .join("");

            const node = createNode(
              "rich-text",
              {
                id: runTransformers(textId),
                original: asString.trim(),
                transformed: marked(asString),
              },
              astPointer
            );
            astPointer.children.push(node);
            collector.length = 0;
            break;
          }
          case "label": {
            let labelId = "";
            for (;;) {
              traverse.next();
              if (!traverse.value()) break;
              if (traverse.value() === ":") break;
              labelId += traverse.value();
            }
            let labelValue = "";
            for (;;) {
              traverse.next();
              if (!traverse.value()) break;
              if (traverse.value() === "\n") break;
              labelValue += traverse.value();
            }

            let transformedLabelValue = runTransformers(labelValue);
            if (
              typeof transformedLabelValue != "object" &&
              typeof transformedLabelValue === "string"
            ) {
              transformedLabelValue = {
                type: "text",
                value: transformedLabelValue,
              };
            }
            const labelNode = createNode(
              "label",
              {
                id: runTransformers(labelId),
                value: transformedLabelValue,
              },
              astPointer
            );
            collector.length = 0;
            astPointer.children.push(labelNode);
            break;
          }
        }
      } else {
        if (traverse.value().trim() === "") continue;
        collector.push(traverse.value());
      }
    }
    return ast;
  } catch (_err) {
    return ast;
  }
}

function runTransformers(literal: string | Record<string, unknown>) {
  let acc = literal;
  for (const tranform of DEFAULT_TRANSFORMS) {
    if (typeof acc === "object" && acc.type) return acc;
    if (typeof acc === "string") acc = tranform(acc);
  }
  return acc;
}

function dateTranform(code: string) {
  const trimmedCode = code.trim();
  if (!trimmedCode.startsWith("date")) {
    return code;
  }

  const dateLiteral = trimmedCode.slice("date".length).trim();
  return createNode("date", new Date(dateLiteral));
}

function urlTransfrom(code: string) {
  const trimmedCode = code.trim();
  if (!trimmedCode.startsWith("url")) {
    return code;
  }

  const collector = [];
  let scopeDelim;
  const traverser = createTokenTraverser(
    trimmedCode.slice("url".length).split("")
  );

  const literals = [];
  while (traverser.next()) {
    if (/\s+/.test(traverser.value())) {
      if (!scopeDelim) {
        if (collector.length > 0) {
          literals.push(collector.join(""));
          collector.length = 0;
          continue;
        }
      }
    }

    if (traverser.value() === '"') {
      if (collector.length > 0 && scopeDelim === '"') {
        literals.push(collector.join(""));
        collector.length = 0;
        scopeDelim = "";
      } else {
        scopeDelim = '"';
      }
      continue;
    }
    collector.push(traverser.value());
  }

  if (collector.length > 0) {
    literals.push(collector.join(""));
  }

  if (literals.length == 2) {
    const alias = literals[0].trim();
    const value = literals[1].trim();
    return createNode("url", {
      alias: alias,
      link: value,
    });
  } else if (literals.length == 1) {
    const value = literals[0].trim();
    return createNode("url", {
      alias: value,
      link: value,
    });
  }
  return code;
}

function nestedStringLiteralTransform(code: string) {
  const trimmedCode = code.trim();
  if (!(trimmedCode.startsWith('"') && trimmedCode.endsWith('"'))) {
    return trimmedCode;
  }
  return trimmedCode.slice(1, -1);
}

export function parse(code: string, options: ParserOptions): AST {
  const ast = toAst(tokenizer(code), options);
  return ast;
}
