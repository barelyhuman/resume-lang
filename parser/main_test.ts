import { assertSnapshot } from "@std/testing/snapshot";
import { parse } from "./main.ts";

// removes the circular dep on the `parent`
// nodes and creates a simpler Object
function toJSON(ast: any) {
  return JSON.parse(
    JSON.stringify(ast, (k, v) => (k === "parent" ? undefined : v), 2)
  );
}

Deno.test("parse isolated labels", async (t) => {
  const output = parse(
    `
label :"Siddharth Gelera"
label E-Mail: url ahoy@barelyhuman.dev mailto:ahoy@barelyhuman.dev
label Website: url https://reaper.is`,
    {
      rootDir: ".",
      readFile: (_path) => ``,
    }
  );
  await assertSnapshot(t, toJSON(output));
});

Deno.test("parse basic section", async (t) => {
  const output = parse(
    `section Basic
    label :"Siddharth Gelera"
    label E-Mail: url ahoy@barelyhuman.dev mailto:ahoy@barelyhuman.dev
    label Website: url https://reaper.is
end`,
    {
      rootDir: ".",
      readFile: (_path) => ``,
    }
  );
  await assertSnapshot(t, toJSON(output));
});

Deno.test("parse nested sections", async (t) => {
  const output = parse(
    `section Education
    section College
      label "Grade":"1"
      label "BatchOf":"2013"
    end
end`,
    {
      rootDir: ".",
      readFile: (_path) => ``,
    }
  );
  await assertSnapshot(t, toJSON(output));
});

Deno.test("imports", async (t) => {
  const fileOne = `section Education
  section College
    label "Grade":"1"
    label "BatchOf":"2013"
  end
end`;
  const fileTwo = `label x:y`;
  const output = parse(
    `
@import "./fileOne"
@import "./fileTwo"
`,
    {
      rootDir: ".",
      readFile: (path) => {
        return path === "./fileOne.resume" ? fileOne : fileTwo;
      },
    }
  );
  await assertSnapshot(t, toJSON(output));
});

Deno.test("rich-text", async (t) => {
  const output = parse(
    `
text Desc:
# heading

some *italics*, some **bold** and 
- a list
end
`,
    {
      rootDir: ".",
      readFile: (_path) => {
        return ``;
      },
    }
  );
  await assertSnapshot(t, toJSON(output));
});

Deno.test("rich-text conflicts", async (t) => {
  const output = parse(
    `
text Desc:
at some point everything will end
end

text Desc:
at some point everything will 
end.
end

text :
at some point everything will 
end.
end

section Section
  text innerText: at some point everything will end
  end
end
`,
    {
      rootDir: ".",
      readFile: (_path) => {
        return ``;
      },
    }
  );
  await assertSnapshot(t, toJSON(output));
});