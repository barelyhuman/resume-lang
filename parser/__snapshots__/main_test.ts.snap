export const snapshot = {};

snapshot[`parse isolated labels 1`] = `
{
  children: [
    {
      children: [],
      type: "label",
      value: {
        id: "",
        value: {
          type: "text",
          value: "Siddharth Gelera",
        },
      },
    },
    {
      children: [],
      type: "label",
      value: {
        id: "E-Mail",
        value: {
          children: [],
          type: "url",
          value: {
            alias: "ahoy@barelyhuman.dev",
            link: "mailto:ahoy@barelyhuman.dev",
          },
        },
      },
    },
    {
      children: [],
      type: "label",
      value: {
        id: "Website",
        value: {
          children: [],
          type: "url",
          value: {
            alias: "https://reaper.is",
            link: "https://reaper.is",
          },
        },
      },
    },
  ],
  type: "root",
}
`;

snapshot[`parse basic section 1`] = `
{
  children: [
    {
      children: [
        {
          children: [],
          type: "label",
          value: {
            id: "",
            value: {
              type: "text",
              value: "Siddharth Gelera",
            },
          },
        },
        {
          children: [],
          type: "label",
          value: {
            id: "E-Mail",
            value: {
              children: [],
              type: "url",
              value: {
                alias: "ahoy@barelyhuman.dev",
                link: "mailto:ahoy@barelyhuman.dev",
              },
            },
          },
        },
        {
          children: [],
          type: "label",
          value: {
            id: "Website",
            value: {
              children: [],
              type: "url",
              value: {
                alias: "https://reaper.is",
                link: "https://reaper.is",
              },
            },
          },
        },
      ],
      type: "section",
      value: "Basic",
    },
  ],
  type: "root",
}
`;

snapshot[`parse nested sections 1`] = `
{
  children: [
    {
      children: [
        {
          children: [
            {
              children: [],
              type: "label",
              value: {
                id: "Grade",
                value: {
                  type: "text",
                  value: "1",
                },
              },
            },
            {
              children: [],
              type: "label",
              value: {
                id: "BatchOf",
                value: {
                  type: "text",
                  value: "2013",
                },
              },
            },
          ],
          type: "section",
          value: "College",
        },
      ],
      type: "section",
      value: "Education",
    },
  ],
  type: "root",
}
`;

snapshot[`imports 1`] = `
{
  children: [
    {
      children: [
        {
          children: [
            {
              children: [],
              type: "label",
              value: {
                id: "Grade",
                value: {
                  type: "text",
                  value: "1",
                },
              },
            },
            {
              children: [],
              type: "label",
              value: {
                id: "BatchOf",
                value: {
                  type: "text",
                  value: "2013",
                },
              },
            },
          ],
          type: "section",
          value: "College",
        },
      ],
      type: "section",
      value: "Education",
    },
    {
      children: [],
      type: "label",
      value: {
        id: "x",
        value: {
          type: "text",
          value: "y",
        },
      },
    },
  ],
  type: "root",
}
`;

snapshot[`rich-text 1`] = `
{
  children: [
    {
      children: [],
      type: "rich-text",
      value: {
        id: "Desc",
        original: "# heading

some *italics*, some **bold** and 
- a list",
        transformed: "<h1>heading</h1>
<p>some <em>italics</em>, some <strong>bold</strong> and </p>
<ul>
<li>a list</li>
</ul>
",
      },
    },
  ],
  type: "root",
}
`;

snapshot[`rich-text conflicts 1`] = `
{
  children: [
    {
      children: [],
      type: "rich-text",
      value: {
        id: "Desc",
        original: "at some point everything will end",
        transformed: "<p>at some point everything will end</p>
",
      },
    },
    {
      children: [],
      type: "rich-text",
      value: {
        id: "Desc",
        original: "at some point everything will 
end.",
        transformed: "<p>at some point everything will 
end.</p>
",
      },
    },
    {
      children: [],
      type: "rich-text",
      value: {
        id: "",
        original: "at some point everything will 
end.",
        transformed: "<p>at some point everything will 
end.</p>
",
      },
    },
    {
      children: [
        {
          children: [],
          type: "rich-text",
          value: {
            id: "innerText",
            original: "at some point everything will end",
            transformed: "<p> at some point everything will end
  </p>
",
          },
        },
      ],
      type: "section",
      value: "Section",
    },
  ],
  type: "root",
}
`;
